activemodel/lib/active_model/type/helpers/time_value.rb

```diff
@@ -65,21 +65,24 @@ def new_time(year, mon, mday, hour, min, sec, microsec, offset = nil)
             end
           end
 
-          ISO_DATETIME = /\A(\d{4})-(\d\d)-(\d\d) (\d\d):(\d\d):(\d\d)(\.\d+)?\z/
+          ISO_DATETIME = /\A(\d{4})-(\d\d)-(\d\d) (\d\d):(\d\d):(\d\d)(\.\d+)?([+|-]\d\d)?\z/
```

Comment:
±hh:mm and ±hhmm are also valid timezone offset.

https://en.wikipedia.org/wiki/ISO_8601#Time_zone_designators

```suggestion
          ISO_DATETIME = /\A(\d{4})-(\d\d)-(\d\d) (\d\d):(\d\d):(\d\d)(\.\d+)?([+-]\d\d(?::?00)?)?\z/
```

Reply:
Here's the problem: adding a regexp for the `±hh:mm and ±hhmm` case slows down the cases without a time zone:

```ruby
require 'active_record'
require 'active_support'
require 'active_record/connection_adapters/postgresql_adapter'
require 'benchmark/ips'

module ActiveRecord
  module ConnectionAdapters
    module PostgreSQL
      module OID # :nodoc:
        class DateTimeFast < DateTime # :nodoc:
          ISO_DATETIME_TZ = /\A(\d{4})-(\d\d)-(\d\d) (\d\d):(\d\d):(\d\d)(\.\d+)?([+|-](\d\d):?(\d\d)?)?\z/

          private

          def fast_string_to_time(string)
            if string =~ ISO_DATETIME_TZ
              microsec = parse_microseconds($7)
              offset = parse_offset($8, $9, $10)
              new_time $1.to_i, $2.to_i, $3.to_i, $4.to_i, $5.to_i, $6.to_i, microsec, offset
            end
          end

          def parse_microseconds(microsec_part)
            if microsec_part && microsec_part.start_with?(".") && microsec_part.length == 7
              microsec_part[0] = ""
              return microsec_part.to_i
            else
              (microsec_part.to_r * 1_000_000).to_i
            end
          end

          def parse_offset(offset_part, hour_part, minutes_part)
            if offset_part
              offset = hour_part.to_i * 3600
              offset += minutes_part.to_i * 60 if minutes_part
            else
              nil
            end
          end

        end
      end
    end
  end
end

timestamps = ['2016-01-02 12:34:56.123456+06:30',
              '2016-01-02 12:34:56.123456+0630',
              '2016-01-02 12:34:56.123456+06',
              '2016-01-02 12:34:56.123456',
              '2016-01-02 12:34:56']

timestamps.each do |timestamp|
  Benchmark.ips do |x|
    x.report("orig") {
      ActiveRecord::ConnectionAdapters::PostgreSQL::OID::DateTime.new.send(:cast_value, timestamp)
    }

    x.report("new") {
      ActiveRecord::ConnectionAdapters::PostgreSQL::OID::DateTimeFast.new.send(:cast_value, timestamp)
    }

    x.compare!
  end
end
```

```
Warming up --------------------------------------
                orig     3.657k i/100ms
                 new    12.309k i/100ms
Calculating -------------------------------------
                orig     40.799k (± 3.4%) i/s -    204.792k in   5.025565s
                 new    128.161k (± 3.0%) i/s -    652.377k in   5.095233s

Comparison:
                 new:   128160.8 i/s
                orig:    40799.1 i/s - 3.14x  slower

Warming up --------------------------------------
                orig     4.024k i/100ms
                 new    12.290k i/100ms
Calculating -------------------------------------
                orig     40.534k (± 2.7%) i/s -    205.224k in   5.066894s
                 new    126.534k (± 2.6%) i/s -    639.080k in   5.054450s

Comparison:
                 new:   126534.0 i/s
                orig:    40534.1 i/s - 3.12x  slower

Warming up --------------------------------------
                orig     3.943k i/100ms
                 new    12.987k i/100ms
Calculating -------------------------------------
                orig     41.384k (± 2.7%) i/s -    208.979k in   5.053328s
                 new    143.288k (± 2.2%) i/s -    727.272k in   5.078117s

Comparison:
                 new:   143288.1 i/s
                orig:    41384.4 i/s - 3.46x  slower

Warming up --------------------------------------
                orig    14.889k i/100ms
                 new    14.170k i/100ms
Calculating -------------------------------------
                orig    156.663k (± 2.2%) i/s -    789.117k in   5.039521s
                 new    144.307k (± 4.2%) i/s -    722.670k in   5.017391s

Comparison:
                orig:   156662.6 i/s
                 new:   144307.2 i/s - 1.09x  slower

Warming up --------------------------------------
                orig    15.513k i/100ms
                 new    11.978k i/100ms
Calculating -------------------------------------
                orig    156.361k (± 7.9%) i/s -    775.650k in   5.009019s
                 new    153.920k (± 3.0%) i/s -    778.570k in   5.063155s

Comparison:
                orig:   156361.5 i/s
                 new:   153919.6 i/s - same-ish: difference falls within error
```